datacapturing/src/main/java/de/cyface/datacapturing/persistence/CapturedDataWriter.java

```diff
@@ -0,0 +1,153 @@
+                    client.bulkInsert(MeasuringPointsContentProvider.SAMPLE_POINTS_URI,
+                            Arrays.copyOfRange(values, startIndex, endIndex));
+                }
+
+                values = new ContentValues[data.getRotations().size()];
+                for (int i = 0; i < data.getRotations().size(); i++) {
+                    Point3D dataPoint = data.getRotations().get(i);
+                    ContentValues ret = new ContentValues();
+                    ret.put(RotationPointTable.COLUMN_RX, dataPoint.getX());
+                    ret.put(RotationPointTable.COLUMN_RY, dataPoint.getY());
+                    ret.put(RotationPointTable.COLUMN_RZ, dataPoint.getZ());
+                    ret.put(RotationPointTable.COLUMN_IS_SYNCED, 0);
+                    ret.put(RotationPointTable.COLUMN_MEASUREMENT_FK, measurementIdentifier);
+                    ret.put(RotationPointTable.COLUMN_TIME, dataPoint.getTimestamp());
+                    values[i] = ret;
+                }
+                for (int startIndex = 0; startIndex < values.length; startIndex += MAX_SIMULTANEOUS_OPERATIONS) {
+                    int endIndex = Math.min(values.length, startIndex + MAX_SIMULTANEOUS_OPERATIONS);
+                    // BulkInsert is about 80 times faster than insertBatch
+                    client.bulkInsert(MeasuringPointsContentProvider.ROTATION_POINTS_URI,
+                            Arrays.copyOfRange(values, startIndex, endIndex));
+                }
+
+                values = new ContentValues[data.getDirections().size()];
+                for (int i = 0; i < data.getDirections().size(); i++) {
+                    Point3D dataPoint = data.getDirections().get(i);
+                    ContentValues ret = new ContentValues();
+                    ret.put(MagneticValuePointTable.COLUMN_MX, dataPoint.getX());
+                    ret.put(MagneticValuePointTable.COLUMN_MY, dataPoint.getY());
+                    ret.put(MagneticValuePointTable.COLUMN_MZ, dataPoint.getZ());
+                    ret.put(MagneticValuePointTable.COLUMN_IS_SYNCED, 0);
+                    ret.put(MagneticValuePointTable.COLUMN_MEASUREMENT_FK, measurementIdentifier);
+                    ret.put(MagneticValuePointTable.COLUMN_TIME, dataPoint.getTimestamp());
+                    values[i] = ret;
+                }
+                for (int startIndex = 0; startIndex < values.length; startIndex += MAX_SIMULTANEOUS_OPERATIONS) {
+                    int endIndex = Math.min(values.length, startIndex + MAX_SIMULTANEOUS_OPERATIONS);
+                    // BulkInsert is about 80 times faster than insertBatch
+                    client.bulkInsert(MeasuringPointsContentProvider.MAGNETIC_VALUE_POINTS_URI,
+                            Arrays.copyOfRange(values, startIndex, endIndex));
+                }
+
+            } catch (DataCapturingException e) {
+                // TODO: not sure if this is ok according to our SDK API?
```

Comment:
If client == null it should be legit to throw an IllegalStateException, can this actually happen when used properly?